letter-shop.md

```diff
@@ -0,0 +1,232 @@
+
+The plugin used here is specific to XRP, and to the use of on-ledger escrow. Escrow for XRP is described [here](https://ripple.com/build/rippleapi/#transaction-types). Escrow transfers differ from normal transfer in that the recipient doesn't automatically receive the amount of the transfer in their account; the need to produce something in order to claim the funds. During the time between the sender's action of preparing the transfer (creating the escrow), and the time the recipient produces the fulfillment for the transfer's condition, the money is on hold on the ledger. If the recipient doesn't produce the fulfillment in time, the transaction is canceled, and the money goes back into the sender's account.
+
+In the case of Interledger, the fulfillment is always a 32-byte string, and the condition is the sha256 hash of that string.
+SHA256 is a one-way hash function, so if you know `fulfillment`, then it's easy and quick to calculate `condition = sha256(fulfillment)`, but if you only know the condition, if you only have a million years or less, it's near-impossible to find (i.e., guess) a fulfillment for which `condition = sha256(fulfillment)` would hold. And in practice, we tend to use rollback timeouts that are of course much shorter! :) Because the transfer is *locked* until the recipient produces the correct fulfillment, the condition of the transfer is a *hash* of its fulfillment, and the transfer will *time* out after a while if the recipient doesn't produce the fulfillment to claim the funds, we call this type of conditional transfer a *Hash Time Lock Contract (HTLC)*. We call this a contract between sender an recipient, that is enforced by the ledger. If the sender and the recipient would exchange the condition and the fulfillment directly, that would be an off-ledger transaction, and instead of a Hash Time Lock Contract, we would use the more general term Hash Time Lock Agreement, to indicate that the interaction happens without having the ledger as an arbitor.
+
+## Paying for your letter
+Visit http://localhost:8000. You'll see something like:
+```txt
+Please send an Interledger payment to test.crypto.xrp.rrhnXcox5bEmZfJCHzPxajUtwdt772zrCW with condition A6-zI1uIEtjOXMTDoZLtML1xj6YPxBA6yxIQyVh4qhE
+```
+
+Save the following script as `pay.js`:
+```js
+const Plugin = require('ilp-plugin-xrp-escrow')
+const uuid = require('uuid/v4')
+
+const plugin = new Plugin({
+  secret: 'sndb5JDdyWiHZia9zv44zSr2itRy1',
+  account: 'rGtqDAJNTDMLaNNfq1RVYgPT8onFMj19Aj',
+  server: 'wss://s.altnet.rippletest.net:51233',
+  prefix: 'test.crypto.xrp.'
+})
+
+function sendTransfer (obj) {
+  obj.id = uuid()
+  obj.from = plugin.getAccount()
+  // to
+  obj.ledger = plugin.getInfo().prefix
+  // amount
+  obj.ilp = 'AA'
+  obj.noteToSelf = {}
```

Comment:
This isn't needed anymore

Reply:
[draft 5](https://interledger.org/rfcs/0004-ledger-plugin-interface/draft-5.html#class-transfer) still has it as a required field? We mentioned making it optional in https://github.com/interledger/rfcs/issues/289 but never actioned that.

Reply:
I'm talking about the `noteToSelf` and maybe the `custom` fields. There used to be a bug where `ilp-plugin-xrp-escrow` would throw an error if those weren't set but I think that's fixed now

Reply:
Regardless of the behavior of this plugin, and regardless of what we may agree as common sense, I'm reading the current version of IL-RFC-4 (draft 5) doesn't say they are optional, so therefore it would be an incorrect use of the LPI if this tutorial were to leave them out when calling `plugin.sendTransfer`.

Reply:
I guess the difference is, you're looking at what the plugin does, I'm looking at what the spec says. :)

Reply:
https://github.com/interledger/rfcs/pull/304

It wasn't explicit in the field documentation but [this example](https://github.com/interledger/rfcs/blob/master/0004-ledger-plugin-interface/0004-ledger-plugin-interface.md#example-1) shows sending a transfer without those fields set.

Reply:
> IL-RFC-4 (draft 5)

Draft numbers finally come in useful! 🍾